File: go/fixture/go-mini/internal/services/heartbeat_parser.go
package services

import (
	"fmt"
	"strings"

	"example.com/go-mini/internal/models"
)

// ParseHeartbeatLine parses a heartbeat line.
func ParseHeartbeatLine(raw string) (id string, status string, version string, tags []string, err error) { //nolint:revive // TODO
	parts := strings.Split(raw, "|")
	if len(parts) < 3 {
		return "", "", "", nil, fmt.Errorf("bad heartbeat %q", raw)
	}
	id = strings.TrimSpace(parts[0])
	if id == "" || len(id) > 64 {
		return "", "", "", nil, fmt.Errorf("bad id %q", id)
	}
	status = strings.ToUpper(strings.TrimSpace(parts[1]))
	if !models.IsValidStatus(status) {
		return id, "", "", nil, fmt.Errorf("unknown status %q", status)
	}
	version = strings.TrimSpace(parts[2])
	if len(parts) > 3 {
		tags = normalizeTags(strings.Split(parts[3], ","))
	}
	return id, status, version, tags, nil
}

// normalizeTags trims, dedupes and drops region tags with unknown codes.
//
//nolint:gocognit,gocyclo,nestif // TODO
func normalizeTags(in []string) (tags []string) {
outer:
	for _, t := range in {
		t = strings.TrimSpace(t)
		if t != "" {
			for _, e := range tags {
				if e == t {
					continue outer // already added. skip
				}
			}
			if strings.HasPrefix(t, "region:") {
				if len(t) > 7 && (t[7:] == "eu" || t[7:] == "us" || t[7:] == "ap") {
					tags = append(tags, t)
				}
			} else {
				tags = append(tags, t)
			}
		}
	}
	return tags
}

File: py/fixture/py-mini/internal/services/heartbeat_parser.py
"""Parsing of the heartbeat wire line."""

from internal.models.status import is_valid_status


def parse_heartbeat_line(raw: str) -> tuple[str, str, str, list[str] | None, str | None]:
    """Parse a heartbeat line."""
    parts = raw.split("|")
    if len(parts) < 3:
        return "", "", "", None, f'bad heartbeat "{raw}"'
    device_id = parts[0].strip()
    if not device_id or len(device_id) > 64:
        return "", "", "", None, f'bad id "{device_id}"'
    status = parts[1].strip().upper()
    if not is_valid_status(status):
        return device_id, "", "", None, f'unknown status "{status}"'
    version = parts[2].strip()
    tags: list[str] | None = None
    if len(parts) > 3:
        tags = _normalize_tags(parts[3].split(","))
    return device_id, status, version, tags, None


def _normalize_tags(inp: list[str]) -> list[str]:  # TODO
    """Trim, dedupe and drop region tags with unknown codes."""
    tags: list[str] = []
    for raw in inp:
        t = raw.strip()
        if t:
            for e in tags:
                if e == t:
                    break  # already added. skip
            else:
                if t.startswith("region:"):
                    if len(t) > 7 and (t[7:] == "eu" or t[7:] == "us" or t[7:] == "ap"):  # noqa: PLR2004  # TODO
                        tags.append(t)
                else:
                    tags.append(t)
    return tags
